code_relocation: Handle relocation of noinit sections

jhol opened the pull request:
* Adds support for relocation of `noinit` data sections.
* Documentation tidy-ups.

jhol committed:
doc: kernel: Improved reloc text formatting

This patch improves the formatting of the "Additional Configurations"
section of the "Code And Data Relocation" document.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
code_relocation: Handle relocation of noinit sections

The code relocation feature allows code and data section to be located
inside a designated RAM region.

Currently, this feature supports relocation of code (text) and data
including mutable data (data), read-only data (rodata), and
zero-initialized data (bss).

However, relocation of non-initialized data sections was not previously
supported, meaning that any data annotated with the __noinit attribute
could not be relocated into the desired RAM region.

This patch adds a NOINIT memory-type which can be used implicitly or
explictly in the zephyr_code_relocate() CMake function. This causes the
build system to generate additional linker-script section-matching rules.

By the nature of noinit data, no action is required by Zephyr at boot.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

JarmouniA reviewed (commented):
Otherwise, looks good to me.

jhol commented on doc/kernel/code-relocation.rst:
```diff
@@ -74,30 +74,30 @@ for  data copy operations from ROM to required memory type.
-* Multiple regions can also be appended together such as: SRAM2_DATA_BSS.
-  This will place data and bss inside SRAM2.
+* Multiple regions can also be appended together such as:
+  ``SRAM2_DATA_BSS_NOINIT``. This will place data and bss inside ``SRAM2``.
```
I applied your suggestion - just with a different wording.

cfriedt commented:
@jhol - can you revisit the commits and separate the documentation commit from the code commit?

cfriedt commented:
> Are you saying you want the second commit split into two?

I'm asking if you can separate the documentation change from the code change. It looks like there was an accidental change here.

https://github.com/zephyrproject-rtos/zephyr/pull/90405/commits/8732a6e53ec8dbc2e1c013c25fb5c2ffd64a94df

jhol commented:
> Those two commits are fine, but there is some leakage between them at the moment.

That's not really the case. This is from the second commit: https://github.com/zephyrproject-rtos/zephyr/pull/90405/commits/8732a6e53ec8dbc2e1c013c25fb5c2ffd64a94df#diff-803bbd7fb5d082c87796454ce2549596d49c7ba11d3f6cd3fb93b40a35123234

All it does is update the docs in correspondence with the functional change.

The first patch just does basic tidy-up of the text.

If you would still prefer this as 3 commits, I can make the change on Monday.

jhol committed:
doc: kernel: code-relocation: Document NOINIT memory type

The `zephyr_code_relocate` now supports the NOINIT memory type in addition
to DATA, TEXT and BSS. This patch documents its usage.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol commented:
> Normally, I would say a separate documentation commit when adding a new feature. So if 3 that means commits, then 3 it should be.

Now split into 3 commits.

jhol commented:
Is there anything else to address here? Or is it good to merge?
